src/utils/jwt.ts

```diff
@@ -0,0 +1,240 @@
+
+export type Dictionary = {[key: string]: any}
+
+export type DecodedToken = {
+  header: Dictionary;
+  payload: Dictionary;
+}
+
+export interface SignatureVerifier {
+  verify(token: string): Promise<void>;
+}
+
+interface KeyFetcher {
+  fetchPublicKeys(): Promise<{ [key: string]: string }>;
+}
+
+export class UrlKeyFetcher implements KeyFetcher {
+  private publicKeys: { [key: string]: string };
+  private publicKeysExpireAt: number;
+
+  constructor(private clientCertUrl: string, private readonly httpAgent?: Agent) {
+    if (!validator.isURL(clientCertUrl)) {
+      throw new Error(
+        'The provided public client certificate URL is an invalid URL.',
+      );
+    }
+  }
+
+  /**
+   * Fetches the public keys for the Google certs.
+   *
+   * @return A promise fulfilled with public keys for the Google certs.
+   */
+  public fetchPublicKeys(): Promise<{ [key: string]: string }> {
+    if (this.shouldRefresh()) {
+      return this.refresh();
+    }
+    return Promise.resolve(this.publicKeys);
+  }
+
+  private shouldRefresh(): boolean {
+    const publicKeysExist = (typeof this.publicKeys !== 'undefined');
+    const publicKeysExpiredExists = (typeof this.publicKeysExpireAt !== 'undefined');
+    const publicKeysStillValid = (publicKeysExpiredExists && Date.now() < this.publicKeysExpireAt);
+    return !(publicKeysExist && publicKeysStillValid);
+  }
+
+  private refresh(): Promise<{ [key: string]: string }> {
+    const client = new HttpClient();
+    const request: HttpRequestConfig = {
+      method: 'GET',
+      url: this.clientCertUrl,
+      httpAgent: this.httpAgent,
+    };
+    return client.send(request).then((resp) => {
+      if (!resp.isJson() || resp.data.error) {
+        // Treat all non-json messages and messages with an 'error' field as
+        // error responses.
+        throw new HttpError(resp);
+      }
+      if (Object.prototype.hasOwnProperty.call(resp.headers, 'cache-control')) {
+        const cacheControlHeader: string = resp.headers['cache-control'];
+        const parts = cacheControlHeader.split(',');
+        parts.forEach((part) => {
+          const subParts = part.trim().split('=');
+          if (subParts[0] === 'max-age') {
+            const maxAge: number = +subParts[1];
+            this.publicKeysExpireAt = Date.now() + (maxAge * 1000);
+          }
+        });
+      }
+      this.publicKeys = resp.data;
+      return resp.data;
+    }).catch((err) => {
+      if (err instanceof HttpError) {
+        let errorMessage = 'Error fetching public keys for Google certs: ';
+        const resp = err.response;
+        if (resp.isJson() && resp.data.error) {
+          errorMessage += `${resp.data.error}`;
+          if (resp.data.error_description) {
+            errorMessage += ' (' + resp.data.error_description + ')';
+          }
+        } else {
+          errorMessage += `${resp.text}`;
+        }
+        throw new JwtError(JwtErrorCode.INTERNAL_ERROR, errorMessage);
```

Comment:
Does it make sense to throw JwtError from this level? I'd think jsonwebtoken will catch the error and convert it into one of its error types.